store/cachekv/store.go

```diff
@@ -49,21 +31,51 @@ func (store *Store) GetStoreType() types.StoreType {
 	return store.parent.GetStoreType()
 }
 
+// Clone creates a snapshot of the cache store.
+// This is a copy-on-write operation and is very fast because
+// it only performs a shadowed copy.
+func (store *Store) Clone() types.CacheKVStore {
+	store.mtx.Lock()
+	defer store.mtx.Unlock()
+
+	return &Store{
+		cache:  store.cache.Copy(),
+		parent: store.parent,
+	}
+}
+
+// swapCache swap out the internal cache store and leave the current store in a unusable state.
+func (store *Store) swapCache() internal.BTree {
+	store.mtx.Lock()
+	defer store.mtx.Unlock()
+
+	cache := store.cache
+	store.cache = internal.BTree{}
+	return cache
+}
+
+// Restore restores the store cache to a given snapshot.
+func (store *Store) Restore(s types.CacheKVStore) {
+	cache := s.(*Store).swapCache()
+
+	store.mtx.Lock()
+	defer store.mtx.Unlock()
+
+	store.cache = cache
+}
```

Comment:
This now means that calling `a.Restore(b)` invalidates/clears the state of `b`. That's surprising, potentially breaks existing code, and isn't part of the documentation of the Restore method! 😬

Reply:
Yeah, I can make it clearer in the documentation later, that's  most common use case that the snapshot is not reused anymore, to reuse it user can do a clone explicitly.

Reply:
I don't think it's reasonable for a method called Restore to destroy the state of its parameter as a side effect. But, whatever.